thrust/system/cuda/detail/adjacent_difference.h

```diff
@@ -61,376 +62,117 @@ namespace cuda_cub {
-      ITEMS_PER_THREAD            = items_per_thread<Tuning::INPUT_SIZE,
-                                          NOMINAL_4B_ITEMS_PER_THREAD>::value
-    };
-    typedef PtxPolicy<128,
-                      ITEMS_PER_THREAD,
-                      cub::BLOCK_LOAD_WARP_TRANSPOSE,
-                      cub::LOAD_LDG,
-                      cub::BLOCK_STORE_WARP_TRANSPOSE>
-        type;
-  };
-
   template <class InputIt,
             class OutputIt,
-            class Size,
             class BinaryOp>
-  struct AdjacentDifferenceAgent
+  cudaError_t THRUST_RUNTIME_FUNCTION
+  doit_step(void *d_temp_storage,
+            size_t &temp_storage_bytes,
+            InputIt first,
+            OutputIt result,
+            BinaryOp binary_op,
+            std::size_t num_items,
+            cudaStream_t stream,
+            bool debug_sync,
+            thrust::detail::integral_constant<bool, false> /* comparable */)
   {
-    typedef typename iterator_traits<InputIt>::value_type input_type;
-
-    // XXX output type must be result of BinaryOp(input_type,input_type);
-    typedef input_type output_type;
-
-    template<class Arch>
-    struct PtxPlan : Tuning<Arch,input_type>::type
+    if (num_items == 0)
     {
-      typedef Tuning<Arch,input_type> tuning;
-
-      typedef typename core::LoadIterator<PtxPlan, InputIt>::type LoadIt;
-      typedef typename core::BlockLoad<PtxPlan, LoadIt>::type     BlockLoad;
-
-      typedef typename core::BlockStore<PtxPlan, OutputIt, input_type>::type
-          BlockStore;
-
-      typedef cub::BlockAdjacentDifference<input_type,
-                                           PtxPlan::BLOCK_THREADS,
-                                           1,
-                                           1,
-                                           Arch::ver>
-          BlockAdjacentDifference;
-
-      union TempStorage
-      {
-        typename BlockAdjacentDifference::TempStorage discontinuity;
-        typename BlockLoad::TempStorage                load;
-        typename BlockStore::TempStorage               store;
-      }; // union TempStorage
-    }; // struct PtxPlan
-
-    typedef typename core::specialize_plan_msvc10_war<PtxPlan>::type::type ptx_plan;
-
-    typedef typename ptx_plan::LoadIt      LoadIt;
-    typedef typename ptx_plan::BlockLoad   BlockLoad;
-    typedef typename ptx_plan::BlockStore  BlockStore;
-    typedef typename ptx_plan::BlockAdjacentDifference BlockAdjacentDifference;
-    typedef typename ptx_plan::TempStorage TempStorage;
-
+      return cudaSuccess;
+    }
 
-    enum
-    {
-      ITEMS_PER_THREAD = ptx_plan::ITEMS_PER_THREAD,
-      BLOCK_THREADS    = ptx_plan::BLOCK_THREADS,
-      ITEMS_PER_TILE   = ptx_plan::ITEMS_PER_TILE,
-    };
+    const auto uint64_threshold = static_cast<std::size_t>(
+      THRUST_NS_QUALIFIER::detail::integer_traits<
+        THRUST_NS_QUALIFIER::detail::int32_t>::const_max);
 
-    struct impl
-    {
+    constexpr bool in_place = true;
```

Comment:
I believe that the current code is correct. The copy in the API name is opposite to performing a copy. `AlgorithmCopy(in, out)` means that `in` and `out` are not allowed to overlap in any way. If pointers aren't compatible, we can't prove if they alias or not: `AlgorithmCopy(double_ptr, reinterpret_cast<char*>(double_ptr))`. I believe there might be more complicated examples. Current version plays safe and apply optimized `*Copy` version only when it can prove that input and output are not overlapped.

Reply:
After discussing this on slack, it seems like we might want to use a different spelling of `in_place` for this algorithm, since `in_place` typically means "input and output are the same memory". Applying it to cases where "input and output may or may not be the same memory" creates some confusion.

This case seems more analogous to a `restrict` annotation on pointers, which documents that two ranges are non-overlapping, rather than `in_place`, which usually guarantees that two ranges overlap exactly.

Something that emphasizes the lack of overlap rather than suggesting that there is overlap would be clearer. Maybe one of these would be better?

- `in_place` -> `may_alias`
- `in_place` -> `!no_alias`
- `in_place` -> `!restrict_ranges`
- `in_place` -> `!unique_ranges`

Since this is an implementation detail that doesn't affect correctness, we don't need to decide/change the names now, but let's at least add a comment clarifying that `in_place` has an unconventional usage here. Eventually it'd be good to rename the `cub::DispatchAdjacentDifference` implementation to avoid confusion.

Reply:
`may_alias` makes the code much more readable, thanks! I'll go ahead and update cub as well.